src/multiversioning.zig

```diff
@@ -0,0 +1,892 @@
+        return writer.print("{}.{}.{}", .{
+            release_triple.major,
+            release_triple.minor,
+            release_triple.patch,
+        });
+    }
+
+    pub fn max(a: Release, b: Release) Release {
+        if (a.value > b.value) {
+            return a;
+        } else {
+            return b;
+        }
+    }
+};
+
+pub const ReleaseTriple = extern struct {
+    patch: u8,
+    minor: u8,
+    major: u16,
+
+    comptime {
+        assert(@sizeOf(ReleaseTriple) == 4);
+        assert(stdx.no_padding(ReleaseTriple));
+    }
+
+    pub fn parse(string: []const u8) error{InvalidRelease}!ReleaseTriple {
+        var parts = std.mem.splitScalar(u8, string, '.');
+        const major = parts.first();
+        const minor = parts.next() orelse return error.InvalidRelease;
+        const patch = parts.next() orelse return error.InvalidRelease;
+        if (parts.next() != null) return error.InvalidRelease;
+        return .{
+            .major = std.fmt.parseUnsigned(u16, major, 10) catch return error.InvalidRelease,
+            .minor = std.fmt.parseUnsigned(u8, minor, 10) catch return error.InvalidRelease,
+            .patch = std.fmt.parseUnsigned(u8, patch, 10) catch return error.InvalidRelease,
+        };
+    }
+};
+
+test "ReleaseTriple.parse" {
+    const tests = [_]struct {
+        string: []const u8,
+        result: error{InvalidRelease}!ReleaseTriple,
+    }{
+        // Valid:
+        .{ .string = "0.0.1", .result = .{ .major = 0, .minor = 0, .patch = 1 } },
+        .{ .string = "0.1.0", .result = .{ .major = 0, .minor = 1, .patch = 0 } },
+        .{ .string = "1.0.0", .result = .{ .major = 1, .minor = 0, .patch = 0 } },
+
+        // Invalid characters:
+        .{ .string = "v0.0.1", .result = error.InvalidRelease },
+        .{ .string = "0.0.1v", .result = error.InvalidRelease },
+        // Invalid separators:
+        .{ .string = "0.0.0.1", .result = error.InvalidRelease },
+        .{ .string = "0..0.1", .result = error.InvalidRelease },
+        // Overflow (and near-overflow):
+        .{ .string = "0.0.255", .result = .{ .major = 0, .minor = 0, .patch = 255 } },
+        .{ .string = "0.0.256", .result = error.InvalidRelease },
+        .{ .string = "0.255.0", .result = .{ .major = 0, .minor = 255, .patch = 0 } },
+        .{ .string = "0.256.0", .result = error.InvalidRelease },
+        .{ .string = "65535.0.0", .result = .{ .major = 65535, .minor = 0, .patch = 0 } },
+        .{ .string = "65536.0.0", .result = error.InvalidRelease },
+    };
+    for (tests) |t| {
+        try std.testing.expectEqualDeep(ReleaseTriple.parse(t.string), t.result);
+    }
+}
+
+pub const MultiversionHeader = extern struct {
+    // When slicing into the binary:
+    // checksum(section[past_offset..past_offset+past_size]) == past_checksum.
+    // This is then validated when the binary is written to a memfd or similar.
+    // TODO: Might be nicer as an AoS? It's control plane state.
```

Comment:
With an AoS, `vsr_releases_max` could be quite flexible (dynamic based on the `PastReleases.count`).

Though, as it is right now, we are more flexible in adding new fields. Not sure sure which dimension of extensibility is more useful!

Reply:
> Though, as it is right now, we are more flexible in adding new fields. Not sure sure which dimension of extensibility is more useful!

Yes, indeed. I'm leaning towards adding new fields; but in either case, we can ~completely change the schema in future by having a transitional release. That is:

* 0.15.4 uses schema version 1
* 0.15.5 uses schema version 1
* 0.15.6 uses schema version 1
* 0.15.7 uses schema version 1/2
* 0.15.8 uses schema version 2

Pack one would have {0.15.4, 0.15.5, 0.15.6, 0.15.7}
Pack two would have {0.15.7, 0.15.8, ...}

There's overlap, so the upgrade process still works as intended, but two steps are required.